tests/test_pipelines.py

```diff
@@ -88,6 +100,65 @@ def test_datasets(self):
                     butler = butlerTests.makeTestRepo(tempRepo)
                     tester.run(butler, self)
 
+    def test_whole_subset(self):
+        """Test that each pipeline's synonymous subset includes all tasks.
+        """
+        files = glob.glob(os.path.join(self.path, "**", "*.yaml"))
+        for file in files:
+            with self.subTest(file):
+                pipeline = lsst.pipe.base.Pipeline.from_uri(file)
+                subset = self.synonyms[os.path.basename(file)]
+                self.assertEqual(pipeline.subsets[subset], set(pipeline.task_labels),
+                                 msg=f"These tasks are missing from subset '{subset}'.")
+
+    def test_ap_pipe_subsets(self):
+        """Test the unique subsets of ApPipe.
+        """
+        files = glob.glob(os.path.join(self.path, "**", "ApPipe*.yaml"))
+        required_subsets = {"preload", "prompt", "afterburner"}
+        # getRegionTimeFromVisit is part of no subset besides apPipe. This is a
+        # very deliberate exception; see RFC-997.
+        no_subset_wanted = {"getRegionTimeFromVisit"}
+
+        for file in files:
+            with self.subTest(file):
+                pipeline = lsst.pipe.base.Pipeline.from_uri(file)
+                # Do all steps exist?
+                self.assertGreaterEqual(pipeline.subsets.keys(), required_subsets,
+                                        msg="An AP pipeline is missing subsets "
+                                            f"{required_subsets - pipeline.subsets.keys()}.")
+                # Is each task part of exactly one step?
+                for set1, set2 in itertools.product(required_subsets, required_subsets):
+                    if set1 == set2:
+                        continue
+                    tasks1 = pipeline.subsets[set1]
+                    tasks2 = pipeline.subsets[set2]
+                    self.assertTrue(tasks1.isdisjoint(tasks2),
+                                    msg=f"Subsets '{set1}' and '{set2}' share tasks "
+                                        f"{tasks1.intersection(tasks2)}.")
+                subsetted = set().union(*[pipeline.subsets[s] for s in required_subsets])
+                self.assertEqual(subsetted, set(pipeline.task_labels) - no_subset_wanted,
+                                 msg=f"These tasks are not in any of the subsets {required_subsets}.")
+
+    def test_inherited_subsets(self):
+        """Test that instrument-specific pipelines have all the subsets of their
+        generic counterparts.
+
+        Note that this does not check inheritance *within* `_ingredients`!
+        """
+        files = [f for f in glob.glob(os.path.join(self.path, "**", "*.yaml"))
+                 if "_ingredients" not in f]
+        for file in files:
+            with self.subTest(file):
+                generic = os.path.join(self.path, "_ingredients", os.path.basename(file))
+                if not os.path.exists(generic):
+                    continue
```

Comment:
Is it ok that this would miss e.g. `LSSTCam/ApPipeCrowded.yaml` if there was no `_ingredients/ApPipeCrowded.yaml`? I'm wondering whether we'd ever define a new pipeline that way, with a different name but built on e.g. `_ingredients/ApPipe.yaml`? The various ISR definitions in `DECam/` don't count, because they're not based on a generic pipeline.

Reply:
The DECam-only pipeline files are exactly why I added this check.

I think it's very plausible that a future `ApPipeCrowded.yaml` would be LSSTCam-only, because it would involve a lot of fine-tuning of parameters (speaking of, there's a not-yet RFC'd proposal to make `calibrateImage` more auto-tuning so that we don't need a separate pipeline; you might ask Ian about it).

Forcing people to declare a generic `ApPipeCrowded` that doesn't do anything seems like the wrong approach, but I'd rather wait to update the testing strategy until we have a concrete case that needs it. The main thing I'm trying to guard against ATM is people making tweaks (e.g., new analysis tasks) to the existing pipelines.

Reply:
Yeah, that's a good approach.